src/main/java/com/google/googleidentity/oauth2/endpoint/AuthorizationEndpoint.java

```diff
@@ -118,5 +132,67 @@ protected void doPost(HttpServletRequest request, HttpServletResponse response)
         }
     }
 
+    /**
+     * Should be call after all validation in doGet function
+     */
+    private OAuth2Request parseOAuth2RequestFromHttpRequest(HttpServletRequest request) {
+        String responseType = request.getParameter(OAuth2ParameterNames.RESPONSE_TYPE);
+
+        String clientID = request.getParameter(OAuth2ParameterNames.CLIENT_ID);
+
+        ClientDetails client = clientDetailsService.getClientByID(clientID).get();
+
+        String redirectUri = request.getParameter(OAuth2ParameterNames.REDIRECT_URI);
+
+        try {
+            redirectUri = URLDecoder.decode(redirectUri, "utf-8");
+        } catch (UnsupportedEncodingException e) {
+            // We always deal with encoded uri in get request
+        }
+
+        Set<String> scope = OAuth2Utils.parseScope(
+                request.getParameter(OAuth2ParameterNames.SCOPE));
+
+        // Set default scopes
+        if (client.getIsScoped() && scope.isEmpty()) {
+            scope = ImmutableSet.copyOf(client.getScopesList());
+        }
+
+        OAuth2Request.Builder oauth2RequestBuilder =
+                OAuth2Request.newBuilder()
+                        .setRequestAuth(
+                                OAuth2Request.RequestAuth.newBuilder()
+                                        .setClientId(client.getClientId())
+                                        .setUsername(
+                                                OAuth2Utils.getUserSession(request).getUser().get()
+                                                        .getUsername())
+                                        .build())
+                        .setRequestBody(
+                                OAuth2Request.RequestBody.newBuilder()
+                                        .setIsScoped(!scope.isEmpty())
+                                        .addAllScopes(scope)
+                                        .setResponseType(responseType)
+                                        .setRefreshable(responseType.equals("code"))
+                                        .setGrantType(
+                                                responseType.equals("code")
+                                                        ? "authorization_code": "implicit")
+                                        .build());
+
+        if (Strings.isNullOrEmpty(request.getParameter(OAuth2ParameterNames.STATE))) {
+            oauth2RequestBuilder.setAuthorizationResponse(
+                    OAuth2Request.AuthorizationResponse.newBuilder()
+                            .setRedirectUri(redirectUri)
+                            .build());
+        } else {
+            oauth2RequestBuilder.setAuthorizationResponse(
+                    OAuth2Request.AuthorizationResponse.newBuilder()
+                            .setState(request.getParameter(OAuth2ParameterNames.STATE))
+                            .setRedirectUri(redirectUri)
+                            .build());
+        }
```

Comment:
nit: remove duplicate code here. You can also access submessage builders to make things a bit simpler.

```suggestion
        oauth2RequestBuilder.getAuthorizationResponseBuilder()
            .setRedirectUri(redirectUri);
        if (!Strings.isNullOrEmpty(request.getParameter(OAuth2ParameterNames.STATE))) {
            oauth2RequestBuilder.getAuthorizationResponseBuilder()
                .setState(request.getParameter(OAuth2ParameterNames.STATE));
        }
```